### DISCUSSION TITLE
SX1276 APRS Not working as expected

### DISCUSSION THREAD

**Discussion by Ilikemining1** [jgromes/RadioLib#702] [General]
Hi,

I'm using a pair of SX1276 ICs in a design of mine, more or less just a copy of the reference designs.  I have both a 900MHz and 144MHz radio, and the eventual plan is to use the 144MHz radio for APRS.  However, for testing, I was using the 900MHz radio since it's an ISM band where I don't have to worry about following amateur rules.  When transmitting APRS packets, only a very few are able to be decoded by the HackRF I'm using as the receiver, and when testing on the 144MHz radio, none were able to be decoded.  LoRa seems to work fine on the 900MHz radio with no issues communicating between a pair of the same radio.  Do you think this is an issue with the way APRS is implemented using external modulation?  These radios are being used with an RP2040 running the arduino-pico core.

Thanks!

IlikeTech

**Comment by jgromes** [comment]
I can't really tell what could be going wrong, since I'm lacking quite a bit of information:

* Can you post your code?
* What is the receiver chain and how is it set up? Eg. HackRF -> SDR# in FM mode -> soundmodem ...

> Do you think this is an issue with the way APRS is implemented using external modulation

I'm not sure what you're referring to with "external modulation".

**Reply by Ilikemining1** [reply]
Hi, 

Sorry for not providing enough information.  Currently, the receiver is the APRS decoder on the Portapack H1 on the HackRF.  As far as my code goes, it's a lightly modified version of the APRS example code:

```
#include <RadioLib.h>

// SX1278 has the following connections:
// NSS pin:   9
// DIO0 pin:  20
// RESET pin: 7
// DIO1 pin:  21
SX1276 radio = new Module(9, 20, 7, 21);

AFSKClient audio(&radio, 22);

// create AX.25 client instance using the AFSK instance
AX25Client ax25(&audio);

// create APRS client instance using the AX.25 client
APRSClient aprs(&ax25);

void setup() {
  Serial.begin(115200);

  Serial.print(F("[SX1278] Initializing ... "));
  int state = radio.beginFSK(902.3);
  
  radio.setRfSwitchPins(2, 3);

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  // initialize AX.25 client
  Serial.print(F("[AX.25] Initializing ... "));

  state = ax25.begin("N0CAL");
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }

  // initialize APRS client
  Serial.print(F("[APRS] Initializing ... "));
  // symbol:                      '>' (car)
  state = aprs.begin('>');
  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while(true);
  }
}

void loop() {
  Serial.print(F("[APRS] Sending position ... "));
  
  // send a location with message and timestamp
  int state = aprs.sendPosition("N0CAL", 0, "2354.49N", "05631.23W", "Testing!", "075500z");
  delay(500);

  if(state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
  }
  
  delay(5000);
}
```

By external modulation I was just referring to how the AFSK modulation is done outside the the SX1276 with DIO2, as opposed to being done by the transceiver chip itself with data being fed over the SPI bus like in LoRa mode, which seems to work fine.

Thanks!

**Reply by jgromes** [reply]
I tried transmitting at 144 MHz using SX1278 on Arduino Uno and my setup with Airspy, SDR console in NFM and soundmodem decodes the APRS with 100% success rate. Unfortunately this is as close as I can get to your setup.

More generally, I don't think changing the carrier should lead to a change in behavior in the library, sounds more like a problem in the RF chain. The only point that's a bit strange is the RF switch - does it work for both 144 and 902.3 MHz?

**Reply by Ilikemining1** [reply]
I'll try setting up the same setup you have going for APRS.  Sorry, that RF switch directive is actually being used to drive RX/TX LEDs, so it's a bit misleading.  I assume there's nothing wrong with doing that?  I wonder if the issues I'm having might be related to the tone generation functions in the arduino-pico board SDK.

Thanks!

**Reply by jgromes** [reply]
Using RF switch for LED signalling should be fine (it's just a couple of calls to digitalWrite in the end) - creative, I like it :)

>  I wonder if the issues I'm having might be related to the tone generation functions in the arduino-pico board SDK

As I don't have the board, I can't test it. Which Arduino core are you using? The official one (https://github.com/arduino/ArduinoCore-mbed) or the unofficial one (https://github.com/earlephilhower/arduino-pico)? If it's the former, there's a bit of a hack to emulate `tone` using Mbed PWM library. The latter has built-in `tone` support.

**Reply by Ilikemining1** [reply]
Thanks for confirming the RF switch thing.  There is an RF switch for the 900MHz radio but it's driven by the radio IC directly using the RX/TX pin.  As far as the core, I'm using the latter, arduino-pico core.  I think I'll take a scope to the D2 pin, because it seems like the tone is very quiet when I listen to the transmissions with my SDR setup, and it's not able to decode them either.  Thanks!

**Reply by jgromes** [reply]
As a simpler test, you can connect a piezospekaer (if you have one) to the pin output and you should hear the AX.25 frames.
